lib/puppet/daemon.rb

```diff
@@ -96,7 +96,11 @@ def reload
 
   def restart
     Puppet::Application.restart!
-    reexec unless agent and agent.running?
+    while agent and agent.running? do
+       Puppet.debug("Agent already running. Retrying in 1 second")
+       sleep(1)
+    end
+    reexec
```

Comment:
I'm thinking we shouldn't wait indefinitely. Maybe make it hard-coded for now, and then observe the `waitforlock` and `maxwaitforlock` settings when they are added. If we reach the max, I assume log an error and `exit 1`


Maybe make that `Puppet.info`, as surfacing the issue during `puppet agent -t` would be helpful for understand what's happening.

Reply:
I don't understand the purpose of this `running?` check, as between the lock is checked and acquired another puppet can start take and the lock.
I suppose the daemon should always start no matter the lock state as lock will be again checked by agent at each interval run.

Reply:
Ah yes, that makes more sense, always reexec and let the re-exec'ed process try to get the lock, or wait until it becomes available.

Reply:
Also note `Puppet::Daemon` used to support running either `puppet agent` or the `puppet master`. But the latter was removed in puppet 4 or 5. It might be good to change the `Daemon#initialize` method so that the [`agent` object is passed in (instead of using an accessor)](https://github.com/puppetlabs/puppet/blob/69a5d3fc4f18711645cf53db1168d12a2c9c0494/lib/puppet/application/agent.rb#L492-L494). Then the check for `unless agent` can be removed.

Reply:
Since it seems that [5.5.x](https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/application/master.rb#L290) only deprecated (did not completely remove) running `master` in daemon, this PR contains only the bug fix and the cleanup was done in a different [PR](https://github.com/puppetlabs/puppet/pull/8135) (targeted to master).